packages/core/src/link-refs.ts

```diff
@@ -0,0 +1,38 @@
+import { SolcBytecode } from './solc-api';
+
+export interface LinkReference {
+  src: string;
+  name: string;
+  length: number;
+  start: number;
+  placeHolder: string;
+}
+
+export function extractLinkReferences(bytecode: SolcBytecode): LinkReference[] {
+  const linkRefs: LinkReference[] = [];
+  const { linkReferences } = bytecode;
+  for (const source of Object.keys(linkReferences)) {
+    for (const name of Object.keys(linkReferences[source])) {
+      const { length, start } = linkReferences[source][name][0];
+      const placeHolder = bytecode.object.substr(start * 2, length * 2);
+      linkRefs.push({
+        src: source,
+        name,
+        length,
+        start,
+        placeHolder,
+      });
+    }
+  }
+  return linkRefs;
+}
+
+export function replaceLinkReferences(bytecode: string, linkReferences: LinkReference[]): string {
+  let unlinkedBytecode: string = bytecode;
+  for (const linkRef of linkReferences) {
+    const { length, start, placeHolder } = linkRef;
+    unlinkedBytecode =
+      unlinkedBytecode.substr(0, 2 + start * 2) + placeHolder + unlinkedBytecode.substr(2 + (start + length) * 2);
```

Comment:
I feel uneasy about the `2 +` part, especially given that `extractLinkReferences` doesn't have that (presumably because Solidity returns the hex encoding without the `0x` prefix).

I think I would prefer if we removed the `0x` prefix at the beginning of the function ([as here](https://github.com/OpenZeppelin/openzeppelin-upgrades/blob/4f2a046804774cc10e15c989afdfc8a3d4bc176a/packages/core/src/version.ts#L21)), and then re-adding it in the return statement.

Reply:
Yes will do that. I agree its not very intuitive.